File: src/ruby/test/projector_test.rb
# encoding: utf-8

require "test/unit"
require_relative "../lib/geodelta/projector"

class GeoDeltaProjectorTest < Test::Unit::TestCase
  def setup
    @mod = GeoDelta::Projector
  end

  def test_lat_to_my
    assert_in_delta(+1.0, @mod.lat_to_my(+85.0511), 1.0E-5)
    assert_equal(0.0, @mod.lat_to_my(0.0))
    assert_in_delta(-1.0, @mod.lat_to_my(-85.0511), 1.0E-5)
  end

  def test_lng_to_mx
    assert_equal(+1.0, @mod.lng_to_mx(+180.0))
    assert_equal(+0.5, @mod.lng_to_mx( +90.0))
    assert_equal( 0.0, @mod.lng_to_mx(   0.0))
    assert_equal(-0.5, @mod.lng_to_mx( -90.0))
    assert_equal(-1.0, @mod.lng_to_mx(-180.0))
  end

  def test_my_to_lat
    assert_in_delta(+85.0511, @mod.my_to_lat(+1.0), 1.0E-4)
    assert_equal(0.0, @mod.my_to_lat(0.0))
    assert_in_delta(-85.0511, @mod.my_to_lat(-1.0), 1.0E-4)
  end

  def test_mx_to_lng
    assert_equal( -90.0, @mod.mx_to_lng(+1.5))
    assert_equal(-180.0, @mod.mx_to_lng(+1.0))
    assert_equal( +90.0, @mod.mx_to_lng(+0.5))
    assert_equal(   0.0, @mod.mx_to_lng( 0.0))
    assert_equal( -90.0, @mod.mx_to_lng(-0.5))
    assert_equal(-180.0, @mod.mx_to_lng(-1.0))
    assert_equal( +90.0, @mod.mx_to_lng(-1.5))
  end

  def test_my_to_ny
    max = @mod::DELTA_HEIGHT
    assert_equal(+12.0, @mod.my_to_ny(+max))
    assert_equal(  0.0, @mod.my_to_ny( 0.0))
    assert_equal(-12.0, @mod.my_to_ny(-max))
  end

  def test_mx_to_nx
    assert_equal(+12.0, @mod.mx_to_nx(+1.0))
    assert_equal(  0.0, @mod.mx_to_nx( 0.0))
    assert_equal(-12.0, @mod.mx_to_nx(-1.0))
  end

  def test_ny_to_my
    max = @mod::DELTA_HEIGHT
    assert_equal(+max, @mod.ny_to_my(+12.0))
    assert_equal( 0.0, @mod.ny_to_my(  0.0))
    assert_equal(-max, @mod.ny_to_my(-12.0))
  end

  def test_nx_to_mx
    assert_equal(+1.0, @mod.nx_to_mx(+12.0))
    assert_equal( 0.0, @mod.nx_to_mx(  0.0))
    assert_equal(-1.0, @mod.nx_to_mx(-12.0))
  end

  def test_lat_to_ny
    assert_equal(0.0, @mod.lat_to_ny(0.0))
    assert_equal(
      @mod.my_to_ny(@mod.lat_to_my(+82.4674)),
      @mod.lat_to_ny(+82.4674))
    assert_equal(
      @mod.my_to_ny(@mod.lat_to_my(-82.4674)),
      @mod.lat_to_ny(-82.4674))
  end

  def test_lng_to_nx
    assert_equal(0.0, @mod.lng_to_nx(0.0))
    assert_equal(
      @mod.mx_to_nx(@mod.lng_to_mx(+180.0)),
      @mod.lng_to_nx(+180.0))
    assert_equal(
      @mod.mx_to_nx(@mod.lng_to_mx(-180.0)),
      @mod.lng_to_nx(-180.0))
  end

  def test_ny_to_lat
    assert_equal(0.0, @mod.ny_to_lat(0.0))
    assert_equal(
      @mod.my_to_lat(@mod.ny_to_my(+12.0)),
      @mod.ny_to_lat(+12.0))
    assert_equal(
      @mod.my_to_lat(@mod.ny_to_my(-12.0)),
      @mod.ny_to_lat(-12.0))
  end

  def test_nx_to_lng
    assert_equal(0.0, @mod.nx_to_lng(0.0))
    assert_equal(
      @mod.mx_to_lng(@mod.nx_to_mx(+12.0)),
      @mod.nx_to_lng(+12.0))
    assert_equal(
      @mod.mx_to_lng(@mod.nx_to_mx(-12.0)),
      @mod.nx_to_lng(-12.0))
  end

  def test_latlng_to_nxy
    assert_equal([0.0, 0.0], @mod.latlng_to_nxy(0.0, 0.0))
    assert_equal(
      [@mod.lng_to_nx(+180.0), @mod.lat_to_ny(+82.4674)],
      @mod.latlng_to_nxy(+82.4674, +180.0))
    assert_equal(
      [@mod.lng_to_nx(-180.0), @mod.lat_to_ny(-82.4674)],
      @mod.latlng_to_nxy(-82.4674, -180.0))
  end

  def test_nxy_to_latlng
    assert_equal([0.0, 0.0], @mod.nxy_to_latlng(0.0, 0.0))
    assert_equal(
      [@mod.ny_to_lat(+12.0), @mod.nx_to_lng(+12.0)],
      @mod.nxy_to_latlng(+12.0, +12.0))
    assert_equal(
      [@mod.ny_to_lat(-12.0), @mod.nx_to_lng(-12.0)],
      @mod.nxy_to_latlng(-12.0, -12.0))
  end

  def test_rush__latlng_nxy
    1000.times {
      lat1, lng1 = [(rand * 180.0) - 90.0, (rand * 360.0) - 180.0]
      nx, ny     = @mod.latlng_to_nxy(lat1, lng1)
      lat2, lng2 = @mod.nxy_to_latlng(nx, ny)
      assert_in_delta(lat1, lat2, 1.0E-10)
      assert_in_delta(lng1, lng2, 1.0E-13)
    }
  end
end

File: src/js/test/projector_test.js

function test_namespace() {
  assertNotUndefined(geodelta);
  assertNotUndefined(geodelta.projector);
}

function test_latToMy() {
  assertRoughlyEquals(+1.0, geodelta.projector.latToMy(+85.0511), 1E-05);
  assertRoughlyEquals( 0.0, geodelta.projector.latToMy(  0.0000), 1E-05);
  assertRoughlyEquals(-1.0, geodelta.projector.latToMy(-85.0511), 1E-05);
}

function test_lngToMx() {
  assertRoughlyEquals(+1.0, geodelta.projector.lngToMx(+180.0), 1E-05);
  assertRoughlyEquals(+0.5, geodelta.projector.lngToMx( +90.0), 1E-05);
  assertRoughlyEquals( 0.0, geodelta.projector.lngToMx(   0.0), 1E-05);
  assertRoughlyEquals(-0.5, geodelta.projector.lngToMx( -90.0), 1E-05);
  assertRoughlyEquals(-1.0, geodelta.projector.lngToMx(-180.0), 1E-05);
}

function test_myToLat()
{
  assertRoughlyEquals(+85.0511, geodelta.projector.myToLat(+1.0), 1E-04);
  assertRoughlyEquals(0.0,      geodelta.projector.myToLat( 0.0), 1E-15);
  assertRoughlyEquals(-85.0511, geodelta.projector.myToLat(-1.0), 1E-04);
}

function test_mxToLng()
{
  assertRoughlyEquals( -90.0, geodelta.projector.mxToLng(+1.5), 1E-15);
  assertRoughlyEquals(-180.0, geodelta.projector.mxToLng(+1.0), 1E-15);
  assertRoughlyEquals( +90.0, geodelta.projector.mxToLng(+0.5), 1E-15);
  assertRoughlyEquals(   0.0, geodelta.projector.mxToLng( 0.0), 1E-15);
  assertRoughlyEquals( -90.0, geodelta.projector.mxToLng(-0.5), 1E-15);
  assertRoughlyEquals(-180.0, geodelta.projector.mxToLng(-1.0), 1E-15);
  assertRoughlyEquals( +90.0, geodelta.projector.mxToLng(-1.5), 1E-15);
}

function test_myToNy()
{
  var max = geodelta.projector.DELTA_HEIGHT();
  assertRoughlyEquals(+12.0, geodelta.projector.myToNy(+max), 1E-15);
  assertRoughlyEquals(  0.0, geodelta.projector.myToNy( 0.0), 1E-15);
  assertRoughlyEquals(-12.0, geodelta.projector.myToNy(-max), 1E-15);
}

function test_mxToNx()
{
  assertRoughlyEquals(+12.0, geodelta.projector.mxToNx(+1.0), 1E-15);
  assertRoughlyEquals(  0.0, geodelta.projector.mxToNx( 0.0), 1E-15);
  assertRoughlyEquals(-12.0, geodelta.projector.mxToNx(-1.0), 1E-15);
}

function test_nyToMy()
{
  var max = geodelta.projector.DELTA_HEIGHT();
  assertRoughlyEquals(+max, geodelta.projector.nyToMy(+12.0), 1E-15);
  assertRoughlyEquals( 0.0, geodelta.projector.nyToMy(  0.0), 1E-15);
  assertRoughlyEquals(-max, geodelta.projector.nyToMy(-12.0), 1E-15);
}

function test_nxToMx()
{
  assertRoughlyEquals(+1.0, geodelta.projector.nxToMx(+12.0), 1E-15);
  assertRoughlyEquals( 0.0, geodelta.projector.nxToMx(  0.0), 1E-15);
  assertRoughlyEquals(-1.0, geodelta.projector.nxToMx(-12.0), 1E-15);
}

function test_latToNy()
{
  assertRoughlyEquals(0.0, geodelta.projector.latToNy(0.0), 1E-15);
  assertRoughlyEquals(geodelta.projector.myToNy(geodelta.projector.latToMy(+82.4674)), geodelta.projector.latToNy(+82.4674), 1E-15);
  assertRoughlyEquals(geodelta.projector.myToNy(geodelta.projector.latToMy(-82.4674)), geodelta.projector.latToNy(-82.4674), 1E-15);
}

function test_lngToNx()
{
  assertRoughlyEquals(0.0, geodelta.projector.lngToNx(0.0), 1E-15);
  assertRoughlyEquals(geodelta.projector.mxToNx(geodelta.projector.lngToMx(+180.0)), geodelta.projector.lngToNx(+180.0), 1E-15);
  assertRoughlyEquals(geodelta.projector.mxToNx(geodelta.projector.lngToMx(-180.0)), geodelta.projector.lngToNx(-180.0), 1E-15);
}

function test_nyToLat()
{
  assertRoughlyEquals(0.0, geodelta.projector.nyToLat(0.0), 1E-15);
  assertRoughlyEquals(geodelta.projector.myToLat(geodelta.projector.nyToMy(+12.0)), geodelta.projector.nyToLat(+12.0), 1E-15);
  assertRoughlyEquals(geodelta.projector.myToLat(geodelta.projector.nyToMy(-12.0)), geodelta.projector.nyToLat(-12.0), 1E-15);
}

function test_nxToLng()
{
  assertRoughlyEquals(0.0, geodelta.projector.nxToLng(0.0), 1E-15);
  assertRoughlyEquals(geodelta.projector.mxToLng(geodelta.projector.nxToMx(+12.0)), geodelta.projector.nxToLng(+12.0), 1E-15);
  assertRoughlyEquals(geodelta.projector.mxToLng(geodelta.projector.nxToMx(-12.0)), geodelta.projector.nxToLng(-12.0), 1E-15);
}

function test_latLngToNxy()
{
  assertRoughlyEquals(0.0, geodelta.projector.latLngToNxy(0.0, 0.0).nx, 1E-15);
  assertRoughlyEquals(0.0, geodelta.projector.latLngToNxy(0.0, 0.0).ny, 1E-15);
  assertRoughlyEquals(geodelta.projector.lngToNx(+180.0000), geodelta.projector.latLngToNxy(+82.4674, +180.0).nx, 1E-15);
  assertRoughlyEquals(geodelta.projector.latToNy( +82.4674), geodelta.projector.latLngToNxy(+82.4674, +180.0).ny, 1E-15);
  assertRoughlyEquals(geodelta.projector.lngToNx(-180.0000), geodelta.projector.latLngToNxy(-82.4674, -180.0).nx, 1E-15);
  assertRoughlyEquals(geodelta.projector.latToNy( -82.4674), geodelta.projector.latLngToNxy(-82.4674, -180.0).ny, 1E-15);
}

function test_nxyToLatLng()
{
  assertRoughlyEquals(0.0, geodelta.projector.nxyToLatLng(0.0, 0.0).lat, 1E-15);
  assertRoughlyEquals(0.0, geodelta.projector.nxyToLatLng(0.0, 0.0).lng, 1E-15);
  assertRoughlyEquals(geodelta.projector.nyToLat(+12.0), geodelta.projector.nxyToLatLng(+12.0, +12.0).lat, 1E-15);
  assertRoughlyEquals(geodelta.projector.nxToLng(+12.0), geodelta.projector.nxyToLatLng(+12.0, +12.0).lng, 1E-15);
  assertRoughlyEquals(geodelta.projector.nyToLat(-12.0), geodelta.projector.nxyToLatLng(-12.0, -12.0).lat, 1E-15);
  assertRoughlyEquals(geodelta.projector.nxToLng(-12.0), geodelta.projector.nxyToLatLng(-12.0, -12.0).lng, 1E-15);
}

/*
TODO: function test_randomLatLngAndNxy()
{
  final double latMax = geodelta.projector.nyToLat(+12.0);
  final double lngMax = 180.0;
  final Random r = new Random();

  for ( int i = 0; i < 1000; i++ )
  {
    final double lat1 = r.nextDouble() * latMax * 2 - latMax;
    final double lng1 = r.nextDouble() * lngMax * 2 - lngMax;
    final double nx = geodelta.projector.lngToNx(lng1);
    final double ny = geodelta.projector.latToNy(lat1);
    final double lat2 = geodelta.projector.nyToLat(ny);
    final double lng2 = geodelta.projector.nxToLng(nx);
    assertRoughlyEquals(lat1, lat2, 1E-13);
    assertRoughlyEquals(lng1, lng2, 1E-13);
  }
}
*/
